buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/ElasticsearchNode.java

```diff
@@ -463,15 +467,59 @@ private void logProcessInfo(String prefix, ProcessHandle.Info info) {
 
     private void logFileContents(String description, Path from) {
         logger.error("{} `{}`", description, this);
+        long lineCutoff;
         try(Stream<String> lines = Files.lines(from, StandardCharsets.UTF_8)) {
-            lines
-                .map(line -> "  " + line)
-                .forEach(logger::error);
+            lineCutoff = lines.count();
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+        final Map<String, Long> errorsAndWarnings;
+        try(Stream<String> lines = Files.lines(from, StandardCharsets.UTF_8)) {
+            errorsAndWarnings = lines
+                .filter(line -> line.contains("ERROR") || line.contains("WARN"))
+                .map(this::normalizeLogLine)
+                .collect(Collectors.groupingBy(Function.identity(), Collectors.counting()));
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+        lineCutoff -= 50;
+        try (LineNumberReader reader = new LineNumberReader(Files.newBufferedReader(from))) {
+            String line = null;
+            do {
+                line = reader.readLine();
+                if (line == null) {
+                    continue;
+                }
+                String normalizedLine = normalizeLogLine(line);
+                if (errorsAndWarnings.keySet().contains(normalizedLine)) {
+                    logger.lifecycle("  " + line);
+                    if (errorsAndWarnings.get(normalizedLine) != 1) {
+                        logger.lifecycle("    ↑ repeated another " + errorsAndWarnings.get(normalizedLine) + " times ↑");
+                    }
+                    errorsAndWarnings.remove(normalizedLine);
+                }
+                if (reader.getLineNumber() == lineCutoff) {
+                    logger.lifecycle("    ↓ tail of the log " + from +  " ↓");
+                }
+                if (reader.getLineNumber() >= lineCutoff) {
+                    logger.lifecycle("  " + line);
+                }
+            } while (line != null);
         } catch (IOException e) {
             throw new UncheckedIOException(e);
         }
     }
 
+    private String normalizeLogLine(String line) {
+        if (line.contains("ERROR")) {
+            return line.substring(line.indexOf("ERROR"));
```

Comment:
Why do you remove ERROR and WARNING prefixes? I think we should keep the log line intact.

Reply:
This is just for the normalization part. It's meant as a quick way to remove the timestamp from the start of the line. We never print the normalized line, we print the actual line when we first encounter it and it matches on of the normalized lines.

Reply:
I see the result of this method being placed into the errorsAndWarnings structure, and then later printed from that structure. For example, line 484 and then line 490. What am I missing?

Reply:
You're right, I changed it with my last commit to print from there, after I wrote the comment above, but note that we don't remove `ERROR` or `WARN` that's what the line will begin with. We remove anything before them, namely the timestamp that would make each line unique.

Reply:
Timestamps are important. They can help us understand when a particular internal operation is slow.

Reply:
@rjernst
We still have the entire log on the file-system for cases like that. 
We would show the time-stamp of the first or last occurrence here anyhow, 
while we complicate the implementation more to retain and show this information I'm not entirely correct that it will be important enough most of the time to do so. 
There will be cases where one needs to download and look at the unmodified log. I would expect that one if which timings are involved will be such a case and the log we show won't matter so much anyhow as it's missing information - such as how frequently the repeated lines are occurring in the logs.